File: expansion-packs/bmad-research-dev/templates/servers/github/update_issue.py
"""
GitHub Update Issue Server Wrapper
Wraps GitHub CLI (gh issue edit)

Usage:
    from servers.github.update_issue import update_issue
    update_issue(123, state='closed')
"""

import json
import subprocess
import time
from typing import List, Dict, Optional, Literal


def update_issue(
    issue_number: int,
    title: Optional[str] = None,
    body: Optional[str] = None,
    add_labels: Optional[List[str]] = None,
    remove_labels: Optional[List[str]] = None,
    add_assignees: Optional[List[str]] = None,
    remove_assignees: Optional[List[str]] = None,
    milestone: Optional[str] = None,
    state: Optional[Literal['open', 'closed']] = None
) -> Dict:
    """
    Update an existing GitHub issue

    Args:
        issue_number: Issue number to update
        title: New title
        body: New body
        add_labels: Labels to add
        remove_labels: Labels to remove
        add_assignees: Assignees to add
        remove_assignees: Assignees to remove
        milestone: Milestone to set
        state: State to set (open/closed)

    Returns:
        Updated issue details

    Examples:
        # Update issue title
        update_issue(42, title="Updated title")

        # Add labels
        update_issue(42, add_labels=['status:doing', 'priority:p1'])

        # Close issue
        update_issue(42, state='closed', add_labels=['status:done'])
    """
    print(f"[GitHub] Updating issue #{issue_number}")

    # Build command
    cmd = ['gh', 'issue', 'edit', str(issue_number)]

    if title:
        cmd.extend(['--title', title])

    if body:
        cmd.extend(['--body', body])

    if add_labels:
        cmd.extend(['--add-label', ','.join(add_labels)])

    if remove_labels:
        cmd.extend(['--remove-label', ','.join(remove_labels)])

    if add_assignees:
        cmd.extend(['--add-assignee', ','.join(add_assignees)])

    if remove_assignees:
        cmd.extend(['--remove-assignee', ','.join(remove_assignees)])

    if milestone:
        cmd.extend(['--milestone', milestone])

    subprocess.run(cmd, capture_output=True, text=True, check=True)

    # Handle state change separately if needed
    if state == 'closed':
        subprocess.run(
            ['gh', 'issue', 'close', str(issue_number)],
            capture_output=True, text=True, check=True
        )
    elif state == 'open':
        subprocess.run(
            ['gh', 'issue', 'reopen', str(issue_number)],
            capture_output=True, text=True, check=True
        )

    # Get updated issue
    result = subprocess.run(
        ['gh', 'issue', 'view', str(issue_number), '--json', 'number,title,url,state'],
        capture_output=True, text=True, check=True
    )
    issue = json.loads(result.stdout)

    print(f"[GitHub] Updated issue #{issue['number']}")

    return issue


def update_status(
    issue_number: int,
    new_status: Literal['backlog', 'todo', 'doing', 'review', 'done']
) -> Dict:
    """Update issue status (BMAD workflow)"""
    print(f"[GitHub] Moving issue #{issue_number} to {new_status}")

    # Map of status labels
    status_labels = {
        'backlog': 'status:backlog',
        'todo': 'status:todo',
        'doing': 'status:doing',
        'review': 'status:review',
        'done': 'status:done'
    }

    all_status_labels = list(status_labels.values())
    new_label = status_labels[new_status]

    # Remove old status labels and add new one
    kwargs = {
        'remove_labels': all_status_labels,
        'add_labels': [new_label]
    }

    if new_status == 'done':
        kwargs['state'] = 'closed'

    return update_issue(issue_number, **kwargs)


def move_to_backlog(issue_number: int) -> Dict:
    """Move issue to backlog"""
    return update_status(issue_number, 'backlog')


def move_to_todo(issue_number: int) -> Dict:
    """Move issue to todo"""
    return update_status(issue_number, 'todo')


def move_to_doing(issue_number: int, assignee: Optional[str] = None) -> Dict:
    """Move issue to doing (in progress)"""
    kwargs = {
        'remove_labels': ['status:backlog', 'status:todo', 'status:review', 'status:done'],
        'add_labels': ['status:doing']
    }

    if assignee:
        kwargs['add_assignees'] = [assignee]

    return update_issue(issue_number, **kwargs)


def move_to_review(issue_number: int) -> Dict:
    """Move issue to review"""
    return update_status(issue_number, 'review')


def move_to_done(issue_number: int) -> Dict:
    """Move issue to done (close)"""
    return update_status(issue_number, 'done')


def update_priority(
    issue_number: int,
    priority: Literal['p0', 'p1', 'p2', 'p3']
) -> Dict:
    """Update issue priority"""
    priority_labels = ['priority:p0', 'priority:p1', 'priority:p2', 'priority:p3']
    new_priority = f"priority:{priority}"

    return update_issue(
        issue_number,
        remove_labels=priority_labels,
        add_labels=[new_priority]
    )


def assign_issue(issue_number: int, assignees: any) -> Dict:
    """Assign issue to user"""
    assignee_list = assignees if isinstance(assignees, list) else [assignees]

    return update_issue(issue_number, add_assignees=assignee_list)


def add_comment(issue_number: int, comment: str) -> None:
    """Add comment to issue"""
    print(f"[GitHub] Adding comment to issue #{issue_number}")

    subprocess.run(
        ['gh', 'issue', 'comment', str(issue_number), '--body', comment],
        capture_output=True, text=True, check=True
    )

    print(f"[GitHub] Comment added to issue #{issue_number}")


def add_experiment_results(
    issue_number: int,
    experiment_id: str,
    status: Literal['success', 'failure', 'partial'],
    metrics: Optional[Dict[str, float]] = None,
    findings: Optional[str] = None,
    next_steps: Optional[str] = None
) -> None:
    """Add experiment results comment"""
    status_emoji = {
        'success': '✅',
        'failure': '❌',
        'partial': '⚠️'
    }

    comment_parts = [
        f"## Experiment Results: {experiment_id}\n",
        f"**Status:** {status_emoji[status]} {status.upper()}\n"
    ]

    if metrics:
        comment_parts.append("### Metrics")
        for key, value in metrics.items():
            comment_parts.append(f"- **{key}**: {value}")
        comment_parts.append("")

    if findings:
        comment_parts.append(f"### Findings\n{findings}\n")

    if next_steps:
        comment_parts.append(f"### Next Steps\n{next_steps}")

    comment = '\n'.join(comment_parts)

    add_comment(issue_number, comment)


def update_issues(updates: List[Dict]) -> List[Dict]:
    """Update multiple issues in batch"""
    print(f"[GitHub] Updating {len(updates)} issues...")

    updated = []
    for item in updates:
        issue_number = item['issueNumber']
        changes = item['changes']
        issue = update_issue(issue_number, **changes)
        updated.append(issue)

        # Small delay to avoid rate limiting
        time.sleep(0.5)

    print(f"[GitHub] Updated {len(updated)} issues")

    return updated


def bulk_status_update(
    issue_numbers: List[int],
    new_status: Literal['backlog', 'todo', 'doing', 'review', 'done']
) -> List[Dict]:
    """Bulk status update"""
    updates = []
    for num in issue_numbers:
        changes = {
            'remove_labels': ['status:backlog', 'status:todo', 'status:doing', 'status:review', 'status:done'],
            'add_labels': [f"status:{new_status}"]
        }
        if new_status == 'done':
            changes['state'] = 'closed'

        updates.append({
            'issueNumber': num,
            'changes': changes
        })

    return update_issues(updates)

File: expansion-packs/bmad-research-dev/templates/servers/github/update-issue.ts
/**
 * GitHub Update Issue Server Wrapper
 * Wraps GitHub CLI (gh issue edit)
 *
 * Usage:
 *   import { updateIssue } from './servers/github/update-issue.ts'
 *   await updateIssue(123, { state: 'closed' })
 */

interface UpdateOptions {
  title?: string;
  body?: string;
  addLabels?: string[];
  removeLabels?: string[];
  addAssignees?: string[];
  removeAssignees?: string[];
  milestone?: string;
  state?: 'open' | 'closed';
}

interface Issue {
  number: number;
  title: string;
  url: string;
  state: string;
}

/**
 * Update an existing GitHub issue
 *
 * @param issueNumber - Issue number to update
 * @param updates - Fields to update
 * @returns Updated issue details
 *
 * @example
 * // Update issue title
 * await updateIssue(42, {
 *   title: "Updated title"
 * })
 *
 * @example
 * // Add labels
 * await updateIssue(42, {
 *   addLabels: ['status:doing', 'priority:p1']
 * })
 *
 * @example
 * // Close issue
 * await updateIssue(42, {
 *   state: 'closed',
 *   addLabels: ['status:done']
 * })
 */
export async function updateIssue(
  issueNumber: number,
  updates: UpdateOptions
): Promise<Issue> {
  console.log(`[GitHub] Updating issue #${issueNumber}`)

  const cmdParts = [`gh issue edit ${issueNumber}`];

  if (updates.title) {
    cmdParts.push(`--title "${updates.title.replace(/"/g, '\\"')}"`);
  }

  if (updates.body) {
    cmdParts.push(`--body "${updates.body.replace(/"/g, '\\"')}"`);
  }

  if (updates.addLabels && updates.addLabels.length > 0) {
    cmdParts.push(`--add-label "${updates.addLabels.join(',')}"`);
  }

  if (updates.removeLabels && updates.removeLabels.length > 0) {
    cmdParts.push(`--remove-label "${updates.removeLabels.join(',')}"`);
  }

  if (updates.addAssignees && updates.addAssignees.length > 0) {
    cmdParts.push(`--add-assignee "${updates.addAssignees.join(',')}"`);
  }

  if (updates.removeAssignees && updates.removeAssignees.length > 0) {
    cmdParts.push(`--remove-assignee "${updates.removeAssignees.join(',')}"`);
  }

  if (updates.milestone) {
    cmdParts.push(`--milestone "${updates.milestone}"`);
  }

  const cmd = cmdParts.join(' ');

  await globalThis.Bash({ command: cmd });

  // Handle state change separately if needed
  if (updates.state === 'closed') {
    await globalThis.Bash({
      command: `gh issue close ${issueNumber}`
    });
  } else if (updates.state === 'open') {
    await globalThis.Bash({
      command: `gh issue reopen ${issueNumber}`
    });
  }

  // Get updated issue
  const result = await globalThis.Bash({
    command: `gh issue view ${issueNumber} --json number,title,url,state`
  });
  const issue = JSON.parse(result);

  console.log(`[GitHub] Updated issue #${issue.number}`)

  return issue;
}

/**
 * Update issue status (BMAD workflow)
 */
export async function updateStatus(
  issueNumber: number,
  newStatus: 'backlog' | 'todo' | 'doing' | 'review' | 'done'
): Promise<Issue> {
  console.log(`[GitHub] Moving issue #${issueNumber} to ${newStatus}`)

  // Map of status transitions
  const statusLabels = {
    'backlog': 'status:backlog',
    'todo': 'status:todo',
    'doing': 'status:doing',
    'review': 'status:review',
    'done': 'status:done'
  };

  const allStatusLabels = Object.values(statusLabels);
  const newLabel = statusLabels[newStatus];

  // Remove old status labels and add new one
  return updateIssue(issueNumber, {
    removeLabels: allStatusLabels,
    addLabels: [newLabel],
    ...(newStatus === 'done' && { state: 'closed' })
  });
}

/**
 * Move issue to backlog
 */
export async function moveToBacklog(issueNumber: number): Promise<Issue> {
  return updateStatus(issueNumber, 'backlog');
}

/**
 * Move issue to todo
 */
export async function moveToTodo(issueNumber: number): Promise<Issue> {
  return updateStatus(issueNumber, 'todo');
}

/**
 * Move issue to doing (in progress)
 */
export async function moveToDoing(issueNumber: number, assignee?: string): Promise<Issue> {
  const updates: UpdateOptions = {
    removeLabels: ['status:backlog', 'status:todo', 'status:review', 'status:done'],
    addLabels: ['status:doing']
  };

  if (assignee) {
    updates.addAssignees = [assignee];
  }

  return updateIssue(issueNumber, updates);
}

/**
 * Move issue to review
 */
export async function moveToReview(issueNumber: number): Promise<Issue> {
  return updateStatus(issueNumber, 'review');
}

/**
 * Move issue to done (close)
 */
export async function moveToDone(issueNumber: number): Promise<Issue> {
  return updateStatus(issueNumber, 'done');
}

/**
 * Update issue priority
 */
export async function updatePriority(
  issueNumber: number,
  priority: 'p0' | 'p1' | 'p2' | 'p3'
): Promise<Issue> {
  const priorityLabels = ['priority:p0', 'priority:p1', 'priority:p2', 'priority:p3'];
  const newPriority = `priority:${priority}`;

  return updateIssue(issueNumber, {
    removeLabels: priorityLabels,
    addLabels: [newPriority]
  });
}

/**
 * Assign issue to user
 */
export async function assignIssue(
  issueNumber: number,
  assignees: string | string[]
): Promise<Issue> {
  const assigneeList = Array.isArray(assignees) ? assignees : [assignees];

  return updateIssue(issueNumber, {
    addAssignees: assigneeList
  });
}

/**
 * Add comment to issue
 */
export async function addComment(
  issueNumber: number,
  comment: string
): Promise<void> {
  console.log(`[GitHub] Adding comment to issue #${issueNumber}`)

  await globalThis.Bash({
    command: `gh issue comment ${issueNumber} --body "${comment.replace(/"/g, '\\"')}"`
  });

  console.log(`[GitHub] Comment added to issue #${issueNumber}`)
}

/**
 * Add experiment results comment
 */
export async function addExperimentResults(
  issueNumber: number,
  experimentId: string,
  results: {
    status: 'success' | 'failure' | 'partial';
    metrics?: Record<string, number>;
    findings?: string;
    nextSteps?: string;
  }
): Promise<void> {
  const statusEmoji = {
    success: '✅',
    failure: '❌',
    partial: '⚠️'
  };

  let comment = `## Experiment Results: ${experimentId}\n\n`;
  comment += `**Status:** ${statusEmoji[results.status]} ${results.status.toUpperCase()}\n\n`;

  if (results.metrics) {
    comment += `### Metrics\n`;
    Object.entries(results.metrics).forEach(([key, value]) => {
      comment += `- **${key}**: ${value}\n`;
    });
    comment += '\n';
  }

  if (results.findings) {
    comment += `### Findings\n${results.findings}\n\n`;
  }

  if (results.nextSteps) {
    comment += `### Next Steps\n${results.nextSteps}\n`;
  }

  await addComment(issueNumber, comment);
}

/**
 * Update multiple issues in batch
 */
export async function updateIssues(
  updates: Array<{ issueNumber: number; changes: UpdateOptions }>
): Promise<Issue[]> {
  console.log(`[GitHub] Updating ${updates.length} issues...`)

  const updated = [];
  for (const { issueNumber, changes } of updates) {
    const issue = await updateIssue(issueNumber, changes);
    updated.push(issue);

    // Small delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 500));
  }

  console.log(`[GitHub] Updated ${updated.length} issues`)

  return updated;
}

/**
 * Bulk status update
 */
export async function bulkStatusUpdate(
  issueNumbers: number[],
  newStatus: 'backlog' | 'todo' | 'doing' | 'review' | 'done'
): Promise<Issue[]> {
  return updateIssues(
    issueNumbers.map(num => ({
      issueNumber: num,
      changes: {
        removeLabels: ['status:backlog', 'status:todo', 'status:doing', 'status:review', 'status:done'],
        addLabels: [`status:${newStatus}`],
        ...(newStatus === 'done' && { state: 'closed' })
      }
    }))
  );
}
